### PR TITLE
Aks R divides n^floor(lb B)

### PR DESCRIPTION
**Pull request by metakunt**
Hey @tirix . I think I have messed something up big time.

As mentioned in the last pr I've followed the steps but it appears that I did something wrong. I have a dv condition on ph and p, which makes sure that this theorem application
```
93:91,51,52,53,92:aks4d1p7d1  
                   |- (  (  ph
                         /\ A. p e. Prime ( p || R -> p || N ) )
                      -> R
                         ||
                         ( N ^ ( |_ ` ( 2 logb B ) ) ) )
``` 
now fails.
If anyone knows how to pretty easily fix this one I'd be very happy.

### CODE REVIEW

**Comment by tirix** [comment]
If you have such conflicts you can use `cbvralv` to change the variable `p` into some new variable like `q` which does not have DV constraints in `ph` (`cbv` stands for "change bound variable").
For example:
```
d2::breq1             |- ( p = q -> ( p || R <-> q || R ) )
d3::breq1             |- ( p = q -> ( p || N <-> q || N ) )
d1:d2,d3:imbi12d     |- ( p = q -> ( ( p || R -> p || N ) <-> ( q || R -> q || N ) ) )
d4:d1:cbvralv       |- ( A. p e. Prime ( p || R -> p || N ) <-> A. q e. Prime ( q || R -> q || N ) )
15::                |- ( ( ph /\ A. q e. Prime ( q || R -> q || N ) ) -> R || ( N ^ ( |_ ` ( 2 logb B ) ) ) )
qed:d4,15:sylan2b  |- ( ( ph /\ A. p e. Prime ( p || R -> p || N ) ) -> R || ( N ^ ( |_ ` ( 2 logb B ) ) ) )
```

**Comment by metakunt** [comment]
Ok thanks, let me try it.

EDIT: It worked, thank you very much. Proof should be complete and ready for review soon.

**Comment by metakunt** [comment]
Ok everything worked, which I am quite happy with.
Now the question is how to combine my newly proved aks4d1p7 with aks4d1p5
I need to eliminate this hypothesis:
```
    aks4d1p5.5 $e |- ( ( ( ph /\ 1 < ( N gcd R ) )
                         /\ ( R / ( N gcd R ) ) || A )
                      -> -. ( R / ( N gcd R ) ) || A ) $.
```

File: set.mm [hunk 1]
```diff
@@ -3775,6 +3775,13 @@ intermediate steps that are essentially incomprehensible to humans (other
       ( wn notnotb xchbinxr ) BBDABECF $.
   $}
 
+  ${
+    con2bii2.1 $e |- ( ph <-> -. ps ) $.
+    $( A contraposition inference.  (Contributed by ML, 18-Oct-2020.) $)
+    con2bii2 $p |- ( -. ph <-> ps ) $=
+      ( wn con2bii bicomi ) BADABCEF $.
```
Comment (wlammen) [inline set.mm:3782]: The move of this theorem to main is not justified, in my eyes.  It is a very trivial variation of con2bii, for which we require a minimum of 7 (Norm once notified me of this limit) replacements, so the savings in the end make up for the extra space occupied by the theorem itself.  This minimum is not achieved in this case, although a few instances exist (xor3 for example).  Equivalences like con2bii most of the time feed transitive laws like bitri, that already come in lots of flavors supporting commutation of inputs.  Your application is no exception: con2bii2 is input to a sylib, and you simply could have used con2bii in conjunction with sylibr.

If there were enough examples like xor3, where con2bii2 had lead to simpler proofs, our expectation would have been that your pull request performs the minimizations following from this move.

**Review by wlammen — CHANGES_REQUESTED** [review]

File: set.mm (hunk 1 above)
Comment (metakunt) [reply set.mm:3782]: Look you may be right and I see your point of view, however I don't go ahead and look for those theorems. MMJ2 autofills them for me. Given that it's autofilled and the theorem exists I don't see a particular reason not to use it or move it to main. Hell I didn't even know that I have used a "forbidden" theorem, that should in main in the first place. This just breaks my workflow for no good reason.

Norm is dead and metamath is already the most user unfriendly proof assistant. Following tradition and arbitrary restriction to spite current contributors is asinine to me, especially because I already gave multiple suggestions to move the system to a better place.

Please reconsider your decision, as this is just a move that does not occur extra space.

Comment (wlammen) [reply set.mm:3782]: I'm not sure whether I understand you right.

> I don't see a particular reason not to use it or move it to main. ... Following tradition and arbitrary restriction...

This is a question of quality.  Others (maybe not you) want this database in a concise shape.  A short proof free of extras pleases their esthetic mindset.  In a similar sense they don't want the scenery be littered with pointless theorems obscuring the pillars that a math theory is built upon.  Some might find it annoying to scroll through pages of 'Nearby theorems' in order to find the proper solution for their problem.  Or to page endlessly through the database in their editor.

The rules by which this database is maintained were developed over decades.  Sometimes they are a compromise between conflicting aspects.  Our goal to provide an extremely detailed view on mathematics has the nasty consequence of tons of theorems providing exactly these details.  That is where the ambition to at least limit the proliferation of everything possible might stem from.

If something is added to the database, some effort to keep matters tidy is in order, I think.  You don't open a new road without adjusting the traffic signs in the rest of the town, either.

I'm not willing, and in the position, to dismiss established rules just out of a casual request to do so.

Comment (metakunt) [reply set.mm:3782]: Maybe I should rephrase my statement:
The reason I moved it to main is because it autofilled and shortened my proof. The reason I move stuff to main are usually well thought out.
If you look at https://us.metamath.org/mpeuni/ceilcl.html it has currently 15 usages, 8 are mine, one is ceilcld and  the rest are whatever. Your reservation causes me discomfort in proving things efficiently precisely because I didn't move ceilcld in the first place. Now if I shortened proofs you would quickly see that ceilcld would have 8+2+1=11 usages, far more than the 7 usages of ceilcl.

I don't move stuff to main willy nilly. I have real usages for those theorems and I prefer stuff to work.
Current stuff in main is mostly 1 step too much as I typically have to use syl2anc or syl3anc to convert closed form to deduction form.
This is really inconvenient, especially because there is a lot of cruft in main that is kept for no good reason.

I can list hundreds of examples of theorems that can just be deleted/refactored from the database, yet a theorem that autocompleted is apparently too much.

Especially with propositional logic, the way that I look up those theorems is typing in multiple expressions, enter a colon in front of the statement that I want to deduce and press CTRL+U. If a theorem autocompletes for whatever reason I am taking it. I am not checking the statement if that is in the mathbox until I find out that the pipeline failed. Then and only then I am using the theorem from the mathbox.

In that particular case I was lucky that it autocompleted, usually it is way harder if stuff that should autocomplete does not, because then I have to find the theorems in the database.

I am all for using a simpler versioning. For me https://us.metamath.org/mpeuni/eqtrd.html and https://us.metamath.org/mpeuni/eqcomd.html is enough and all the other inferences are bloat, yet they fill at minimum 50 results which should just be 2. Back then I actually tried to figure out which inferences to use for equality until I have figured out to just use the api of eqtrd and eqcomd, even if the proof may be slightly longer.

Are we willing to remove that bloat? https://us.metamath.org/mpeuni/curryax.html for example, does this have to be in main? Or how about this https://us.metamath.org/mpeuni/pm2.13.html? Literally no usages? Or how about this https://us.metamath.org/mpeuni/pm4.25.html?

I don't see any reason why these theorems are better than the theorem that I have moved to main which has 2 usages and in the future will probably have more, as it is a really useful statement.

https://us.metamath.org/mpeuni/pm2.4.html also no usages with an absolute garbage to use api, but apparently it's worth that it clutters main.

I can't improve the database to actually have a convenient api that I would like to use, I also can't improve the database to convert theorems to deduction form and deprecate the closed form versions.
There are still theorems that are dependent on ax-13 that are part of main. How about moving them to its separate mathbox?

The problem is you follow established rules without knowing why. I am all for cleaning up main set.mm yet all requests for me to do so were denied. And if I actually want to use stuff I have to use an unergonomic api because I couldn't even clean it up in the first place.

Look at https://us.metamath.org/mpeuni/logbleb.html vs my version https://us.metamath.org/mpeuni/logblebd.html
Mine is in the mathbox and has currently 6 usages logbleb has 5 usages, of which logblebd and another usage is from me.
Refactoring logblebd and moving it to main would result in 11 usages of logblebd and the removal of the original logbleb, then after the removal you would have a nice api to use the theorem instead of the horrible mess that is logbleb. Are you willing to greenlight the change, or are you just willing to be a hypocritical pedantic prick about it? Because I've already asked for this change before, which would have improved the database bit by bit.

Basically your refusal and stubbornness causes me to use suboptimal paths that if I did I originally would overtake the suboptimal path quite soon. See the examples of ceilcl and logbleb. It not only makes my proof development harder than it should be for no good reason, it also makes me just not want to be part of the community. 

If I can't move the database to a more ergonomic state with theorems that I actually want to use, then what's the point on just piling on cruft on top of cruft.

If we want to limit main to a set of nice to use details we need an architectural change and a rewrite of the theorems to use a minimal api. That would work and would reduce theorem bloat by a lot, but it would also remove a lot of theorems from it. Best example is eqtrd, I am not learning any theorems except a1i, eqcomd and eqtrd to find any inference that I need. If everything were in deduction form only those 2 theorems would be enough to prove any arbitrary length inequality chain.

This is not a casual request, this is a serious request. I don't ask for this because I really want con2bii2 in main. I don't care about it being in main. I want it (and by it I mean theorems that are for whatever reason useful for my proof development to be in main) to actually be in main, well because for the simple fact that there are, you know, useful. I will eventually have to move the negation symbol from one wff to another and i just want it to work, no exceptions. Having to search for a replacement is just too much to ask for such a simple theorem, as it is such a simple inference with one hypothesis, and a conclusion with three symbols.

In this case this theorem is something I expect to be in the database, anything else is just bad design as I otherwise have to literally search the database painstakingly for a replacement.

Comment (wlammen) [reply set.mm:3782]: I'm short of ignoring your post because of content like this:

> ...willing to be a hypocritical pedantic prick...
> ...your refusal and stubbornness...

Please note that I gave both reasons and solutions to avoid con2bii2.
Hopefully finding back to a more factual and objective tone in the discussion I will this time respond.
1. I think your observation that the database is full of outdated and rarely (if at all) used stuff is right.  The strategy is to be conservative in main.  A theorem might be referred to by a link to the proof on the Metamath web site, so it might cause outside havock if you delete it from the database.
This again is a reason to not lightly add more stuff like con2bii2 to main, as it cannot been removed again.  So you are obliged to prove (not simply state) this theorem is indispensible, and will have lots of applications in the future.
If you cannot supply such a proof you can still add it as a copy to your mathbox. Once you have 15 usages found (like with ceilcl) you can easily convince me and others that an addition to main is in order.
2. Some theorems like pm2.4 are listed because they were extracted from standard literature.  Their usefulness is not automatically assumed, here the motivation is to be complete.
3. ax-13 avoidance is a relative new (about 1 year old) topic.  In my opinion the axiom and its applications should not be removed altogether, since theorems like nfsb have their own right to be shown.
4. It is a matter of belief whether some frequently used convenience theorems like the eq* ones you mentioned have a right to exist.  They have an effect on the proof display, as the keep them shorter and allow the human reader to more easily pinpoint the key steps in the proof.

What can we do to improve on the situation? One possibility is to concentrate key theorems in a so to say infrastructure level, and provide the others in subsections "convenience" and "outdated".  I am on your side as far as this goal is concerned.  I think it is limited resources that prevent really huge changes for now.  There is more like automated proving that would allow converting tautological versions to deduction and immediate ones back and forth, where possible, where no initiative is currently visible.

Comment (metakunt) [reply set.mm:3782]: Here's the thing. I am not frustrated with you. Just so that you understand my perspective and my vision.
I am more frustrated in the bloat that we carry around and can't remove. The reason why I push so hard for deduction theorems versions is, if you convert every theorem from natural numbers onwards to deduction form you would trim down the database by a lot. This alone makes me want to add every theorem that I use even just once as a deduction version because I am quite sure I will need the same derivation in a future proof again.

In particular I have found difficulties with finding the "useful" theorems, right now I have generally a good grasp of what theorems to look for. Ideas were tagging theorems and searching for tags in a better fashion. But this would also need some work.

Basically my idea is having a "core" set of just the most primitive theorems and reexport it. Maybe I'll do some work there first so that you have a good picture of it. Right now it's mostly in my head but I think now may be a good time to start.

Anyways, I'll fix the theorem later, but I am confident with the new core library that I am going to develop soon.

I hope you didn't take it too personal, I am just really getting irked about wanting to make stuff better because I just want to have a better workflow, that's it.

**Comment by tirix** [comment]
> Ok everything worked, which I am quite happy with. 
>
Nice!

> Now the question is how to combine my newly proved aks4d1p7 with aks4d1p5 I need to eliminate this hypothesis:
> 
> ```
>     aks4d1p5.5 $e |- ( ( ( ph /\ 1 < ( N gcd R ) )
>                          /\ ( R / ( N gcd R ) ) || A )
>                       -> -. ( R / ( N gcd R ) ) || A ) $.
> ```
So if I understand well, the bit you have is :
`aks4d1p7 |- ( ph -> E. p e. Prime ( p || R /\ -. p || N ) )`

Once I have such an existence information, the good step is usually to use something like `r19.29a`:
```
10::             |- ( ( ( ( ph /\ 1 < ( N gcd R ) /\ p e. Prime ) /\ ( p || R /\ -. p || N ) ) -> -. ( R / ( N gcd R ) ) || A )
20:              |- ( ( ph /\ 1 < ( N gcd R ) ) -> E. p e. Prime ( p || R /\ -. p || N ) )
30:10,20:r19.29a |- ( ( ph /\ 1 < ( N gcd R ) ) -> -. ( R / ( N gcd R ) ) || A )
```
Now we can work with a specific `p` to prove step `10`. This can go along these lines:
```
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> p || ( R / ( N gcd R ) ) )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> p || A )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> -. R || A )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> -. ( R / p ) || A )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> ( R / p ) e. ( 1 ... B ) )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> R <_ ( R / p ) )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> ( R / p ) < R )
|- ( ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) /\ ( R / ( N gcd R ) ) || A ) -> -. R <_ ( R / p ) )
|- ( ( ( ( ( ph /\ 1 < ( N gcd R ) ) /\ p e. Prime ) /\ p || R ) /\ -. p || N ) -> -. ( R / ( N gcd R ) ) || A )
```
(I think each step is obvious, the last step is again a `pm2.65da`)

**Comment by tirix** [comment]
I realize the first step in the proof might be a big step. Here is one way to prove it as a separated theorem:
```
h1::        |- ( ph -> P e. Prime )
h2:       |- ( ph -> M e. NN )
h3:       |- ( ph -> N e. NN )
h4:   |- ( ph -> P || M )
h5:      |- ( ph -> -. P || N )
10::dvdsgcdb            |- ( ( P e. ZZ /\ M e. ZZ /\ N e. ZZ ) -> ( ( P || M /\ P || N ) <-> P || ( M gcd N ) ) )
11::prmnn                 |- ( P e. Prime -> P e. NN )
d19:1,11:syl             |- ( ph -> P e. NN )
d10:d19:nnzd            |- ( ph -> P e. ZZ )
d11:2:nnzd              |- ( ph -> M e. ZZ )
d12:3:nnzd              |- ( ph -> N e. ZZ )
20:d10,d11,d12,10:syl3anc
                       |- ( ph -> ( ( P || M /\ P || N ) <-> P || ( M gcd N ) ) )
50:5:intnand           |- ( ph -> -. ( P || M /\ P || N ) )
60:50,20:mtbid        |- ( ph -> -. P || ( M gcd N ) )
62::gcdnncl             |- ( ( M e. NN /\ N e. NN ) -> ( M gcd N ) e. NN )
d8:2,3,62:syl2anc      |- ( ph -> ( M gcd N ) e. NN )
70:d19,d8,2:nnproddivdvdsd
                    |- ( ph -> ( ( P x. ( M gcd N ) ) || M <-> P || ( M / ( M gcd N ) ) ) )
d14:d8:nnzd           |- ( ph -> ( M gcd N ) e. ZZ )
75::coprm              |- ( ( P e. Prime /\ ( M gcd N ) e. ZZ ) -> ( -. P || ( M gcd N ) <-> ( P gcd ( M gcd N ) ) = 1 ) )
76:75:biimpa          |- ( ( ( P e. Prime /\ ( M gcd N ) e. ZZ ) /\ -. P || ( M gcd N ) ) -> ( P gcd ( M gcd N ) ) = 1 )
d16:1,d14,60,76:syl21anc
                     |- ( ph -> ( P gcd ( M gcd N ) ) = 1 )
77::gcddvds            |- ( ( M e. ZZ /\ N e. ZZ ) -> ( ( M gcd N ) || M /\ ( M gcd N ) || N ) )
78:d11,d12,77:syl2anc
                      |- ( ph -> ( ( M gcd N ) || M /\ ( M gcd N ) || N ) )
d18:78:simpld        |- ( ph -> ( M gcd N ) || M )
80:d10,d14,d11,d16,4,d18:coprmdvds2d
                    |- ( ph -> ( P x. ( M gcd N ) ) || M )
qed:80,70:mpbid    |- ( ph -> P || ( M / ( M gcd N ) ) )

$= ( cgcd co cmul cdvds wbr wcel cn nnzd syl2anc cz wa cprime prmnn syl gcdnncl
   cdiv wn c1 wceq intnand dvdsgcdb syl3anc mtbid coprm biimpa syl21anc gcddvds
   wb simpld coprmdvds2d nnproddivdvdsd mpbid ) ABCDJKZLKCMNBCVBUEKMNABVBCABABU
   AOZBPOEBUBUCZQZAVBACPODPOVBPOFGCDUDRZQZACFQZAVCVBSOZBVBMNZUFZBVBJKUGUHZEVGAB
   CMNZBDMNZTZVJAVNVMIUIABSOCSOZDSOZVOVJUQVEVHADGQZBCDUJUKULVCVITVKVLBVBUMUNUOH
   AVBCMNZVBDMNZAVPVQVSVTTVHVRCDUPRURUSABVBCVDVFFUTVA $.
```

**Review by wlammen — APPROVED** [review]

**Comment by metakunt** [comment]
@tirix Alright thanks, I'll try it out once I have some time. I quickly looked over the sketch, it seems to have lots of good ideas. I should be able to piece it together.

**Review by avekens — APPROVED** [review]

**Review by tirix — APPROVED** [review]
